src/Dialect/ONNX/ElementsAttr/ElementsAttrBuilder.hpp

```diff
@@ -244,10 +247,32 @@ class ElementsAttrBuilder {
   // Constructs a transformer that changes every element to the result of
   // applying the given function to the element.
   template <typename Function = WideNum (*)(WideNum)>
-  static inline Transformer functionTransformer(Function fun) {
-    return [fun = std::move(fun)](llvm::MutableArrayRef<WideNum> data) -> void {
-      for (WideNum &n : data)
-        n = fun(n);
+  static inline Transformer functionTransformer(
+      Function fun, mlir::MLIRContext *ctx) {
+    return [fun = std::move(fun), ctx](
+               llvm::MutableArrayRef<WideNum> data) -> void {
+      std::mutex mtx;
+      size_t beginOffset = 0;
+      auto fetchBatch = [&](size_t threadNumber) {
+        // Each thread fetches the same batch size. The remainder is set in the
+        // threads with small thread number.
+        const std::lock_guard<std::mutex> lock(mtx);
+        size_t batchSize = data.size() / ctx->getNumThreads();
+        size_t batchSizeMod = data.size() % ctx->getNumThreads();
+        if (threadNumber < batchSizeMod)
+          batchSize += 1;
+        auto batchBegin = data.begin() + beginOffset;
+        auto batchEnd = batchBegin + batchSize;
+        beginOffset += batchSize;
+        return llvm::make_range(batchBegin, batchEnd);
+      };
+
+      auto work = [&](size_t threadNumber) {
+        auto batch = fetchBatch(threadNumber);
+        for (WideNum &n : batch)
+          n = fun(n);
+      };
+      parallelFor(ctx, 0, ctx->getNumThreads(), work);
```

Comment:
As mentioned before, please check that there is enough work to go to parallel computations. I suspect that if the reduction is very small, then we really want to do it sequentially and it will be faster.